docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,113 @@
+| Manufacturer | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Description | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Model | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Labels | string array | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include adding, removing or modifying any of the strings in the array. |
+| Device Resources | array | Situational | New device resources can always be added. Existing device resources can only be removed from the array when the device resource has not yet been associated to any reading or event.|
+| Device Commands | array | Situational | New device commands can always be added.  Existing device commands can only be removed from the array when the device command has not yet been associated to a reading or event.  Please see the device command properties for rules on modifying device commands.Device commands are optional and so the array can always be empty. |
+| *Device Resource Properties* |||
+| Description | string | Anytime | As this property does not affect any functionality, this property can be changed at any time – to include setting the string to an empty string. |
+| Name | string | Situational | When the device resource is not associated to any reading or event, this field can be modified.  It can never be blank and must be unique across the entire profile. |
+| IsHidden | bool | Anytime | This field allows the device resource to be exposed (default) via the command service.  This change does not affect the behavior but only visibility.  This field is optional but is false by default when not provided. |
+| Tag | string | Anytime | As this property does not affect any functionality, this property can be changed at any time – to include setting the string to an empty string. Note: adding or removing a tag could effect queries; that is the same query by tag could return different results if a tag is added or removed between query calls.|
+| Properties | Resource Properties | Situational | When the device resource is not associated to any reading or event, this field can be modified.  Because the ValueType and ReadWrite properties of this field are required, it can never be empty. |
+| Attributes | map | Anytime | The attributes have no impact on existing readings, events, etc.  They only impact how readings are acquired by the service.|
```

Comment:
Attributes is the representation of the Device Resource, such as an address.
When users query a Reading, they might check Profile to see what this Device Resource is referring to.
Thus, it should not be changed if there is any Device associated to this Profile.

Reply:
@lenny-intel , @cloudxxx8, @tonyespy  - we can't say that device profiles can change even when associated to an event or reading and then say attributes can't change.  Which is it???

During the last core working group call, there was agreement that because the event/reading contained the profile name, resource name and even the value type (and because events/readings were short lived) we would allow device profile changes even when associated to an event/reading.  

This note from Cloud would contradict that.  It would say that once attached to a event/reading, a user may want to use the association to that specific device profile with its specific device resources to get additional attribute details.  Therefore, it can't change.

Which is it?  Allow profiles to change regardless of event/reading association or don't allow them because of the need to get at other device profile associated sub objects???

Reply:
@jpwhitemn I didn't attend the last Core WG meeting due to a conflict.

> 
During the last core working group call, there was agreement that because the event/reading contained the profile name, resource name and even the value type (and because events/readings were short lived) we would allow device profile changes even when associated to an event/reading.

Really (hopefully this wasn't based on my comments)? That seems like a big about face, and is in fact contrary to the wording at the top of this ADR which states:

`Allow changes to device profile or device resource fields when the field is inconsequential to the object's use (ex: description) or when it is not yet associated to another EdgeX object. These changes are listed as Anytime or Situational accordingly in the table below.`

I think if we do make this about face, that we ought to consider adding units to the readings too.